### PR TITLE
Extend and omnify charged particle selection

### PR DESCRIPTION
**Pull request by ruse-traveler**
### Briefly, what does this PR introduce?

This PR extends the functionality of the `ChargedParticleSelector` algorithm to also work on `edm4eic::ReconstructedParticle` input and updates the corresponding factories to `JOmniFactory`. This is accomplished by creating a parallel algorithm (`ChargedReconstructedParticleSelector`) and renaming the old algorithm to reflect its scope (`ChargedMCParticleSelector`).

### What kind of change does this PR introduce?
- [] Bug fix (issue #__)
- [X] New feature (issue #1216 )
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [ ] Tests for the changes have been added
- [ ] Documentation has been added / updated
- [X] Changes have been communicated to collaborators

### Does this PR introduce breaking changes? What changes might users need to make to their code?

No.

### Does this PR change default behavior?

No.

### CODE REVIEW

**Comment by ruse-traveler** [comment]
Tagging @nathanwbrei here: a more elegant approach might be to template the old algorithm, but I'm not completely sure how to go about that...

**Comment by wdconinc** [comment]
> template the old algorithm

If you do this, please first verify that this doesn't lead to any problems with http://github.com/eic/algorithms. This may mean we need to adapt to the algorithms interface first to ensure that.

**Comment by ruse-traveler** [comment]
> > template the old algorithm
> 
> If you do this, please first verify that this doesn't lead to any problems with [http://github.com/eic/algorithms](https://github.com/eic/algorithms). This may mean we need to adapt to the algorithms interface first to ensure that.

Gotcha! Personally, I'm leaning towards the PR without templates just to be safe and keep the jet PODIO fix rolling. But this algorithm could be an easy target to test that out!

File: src/algorithms/reco/ChargedMCParticleSelector.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+// Copyright (C) 2023 Dmitry Kalinkin
```
Comment (veprbl) [inline src/algorithms/reco/ChargedMCParticleSelector.h:2]:
```suggestion
// Copyright (C) 2023, 2024, Dmitry Kalinkin, Derek Anderson
```

File: src/algorithms/reco/ChargedReconstructedParticleSelector.h [hunk 2]
```diff
@@ -0,0 +1,45 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+// Copyright (C) 2023 Dmitry Kalinkin
```
Comment (veprbl) [inline src/algorithms/reco/ChargedReconstructedParticleSelector.h:2]:
```suggestion
// Copyright (C) 2023, 2024, Dmitry Kalinkin, Derek Anderson
```

File: src/global/reco/ChargedMCParticleSelector_factory.h [hunk 3]
```diff
@@ -0,0 +1,41 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+// Copyright (C) 2023 Wouter Deconinck, Dmitry Kalinkin
```
Comment (veprbl) [inline src/global/reco/ChargedMCParticleSelector_factory.h:2]:
```suggestion
// Copyright (C) 2023, 2024, Wouter Deconinck, Dmitry Kalinkin, Derek Anderson
```

File: src/global/reco/ChargedReconstructedParticleSelector_factory.h [hunk 4]
```diff
@@ -0,0 +1,40 @@
+/* COPYRIGHT STUFF */
```
Comment (veprbl) [inline src/global/reco/ChargedReconstructedParticleSelector_factory.h:1]:
```suggestion
// Copyright (C) 2024, Derek Anderson
```

File: src/global/reco/ChargedReconstructedParticleSelector_factory.h [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+      PodioInput<edm4eic::ReconstructedParticle> m_pars_in {this, "GeneratedParticles"};
+
+      // output collection
+      PodioOutput<edm4eic::ReconstructedParticle> m_pars_out {this};
+
+    public:
+
+      void Configure() {
+        m_algo = std::make_unique<eicrecon::ChargedReconstructedParticleSelector>();
+        m_algo -> init(logger());
```
Comment (veprbl) [inline src/global/reco/ChargedReconstructedParticleSelector_factory.h:27]:
```suggestion
        m_algo->init(logger());
```

File: src/global/reco/ChargedReconstructedParticleSelector_factory.h [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+      }
+
+      void ChangeRun(int64_t run_number) {
+        /* nothing to do */
+      }
+
+     void Process(int64_t run_number, int64_t event_number) {
+        m_pars_out() = m_algo -> process(m_pars_in());
```
Comment (veprbl) [inline src/global/reco/ChargedReconstructedParticleSelector_factory.h:35]:
```suggestion
        m_pars_out() = m_algo->process(m_pars_in());
```

File: src/global/reco/ChargedMCParticleSelector_factory.h [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+      PodioInput<edm4hep::MCParticle> m_pars_in {this, "GeneratedParticles"};
+
+      // output collection
+      PodioOutput<edm4hep::MCParticle> m_pars_out {this};
+
+    public:
+
+      void Configure() {
+        m_algo = std::make_unique<eicrecon::ChargedMCParticleSelector>();
+        m_algo -> init(logger());
```
Comment (veprbl) [inline src/global/reco/ChargedMCParticleSelector_factory.h:28]:
```suggestion
        m_algo->init(logger());
```

File: src/global/reco/ChargedMCParticleSelector_factory.h [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+      }
+
+      void ChangeRun(int64_t run_number) {
+        /* nothing to do */
+      }
+
+     void Process(int64_t run_number, int64_t event_number) {
+        m_pars_out() = m_algo -> process(m_pars_in());
```
Comment (veprbl) [inline src/global/reco/ChargedMCParticleSelector_factory.h:36]:
```suggestion
        m_pars_out() = m_algo->process(m_pars_in());
```

File: src/global/reco/reco.cc [hunk 9]
```diff
@@ -19,7 +20,6 @@
 #include "MatchClusters_factory.h"
 #include "ReconstructedElectrons_factory.h"
 #include "ReconstructedJets_factory.h"
-#include "algorithms/reco/ChargedParticleSelector.h"
 #include "extensions/jana/JChainMultifactoryGeneratorT.h"
 #include "extensions/jana/JOmniFactoryGeneratorT.h"
```
Comment (veprbl) [inline src/global/reco/reco.cc:24]:
```suggestion
#include "extensions/jana/JOmniFactoryGeneratorT.h"
#include "extensions/spdlog/SpdlogExtensions.h"
```
suggested in https://github.com/eic/EICrecon/actions/runs/7506420354/job/20437970525?pr=1217

File: src/global/reco/reco.cc [hunk 10]
```diff
@@ -19,9 +20,9 @@
...
+#include "extensions/spdlog/SpdlgExtensions.h"
```
Comment (veprbl) [inline src/global/reco/reco.cc:25]:
```suggestion
#include "extensions/spdlog/SpdlogExtensions.h"
```

Comment (ruse-traveler) [reply src/global/reco/reco.cc]: Whoops! Good catch! 😅

File: src/algorithms/reco/ChargedMCParticleSelector.h (hunk 1 above)
Comment (wdconinc) [reply src/algorithms/reco/ChargedMCParticleSelector.h]: Something tells me there is a pre-commit checker for spdx copyright lines...

**Review by veprbl — APPROVED** [review]
LGTM

**Pull request merged**